roadmap/implementers-guide/src/node/approval/approval-voting.md

```diff
@@ -0,0 +1,234 @@
+# Approval Voting
+
+## Protocol
+
+Input:
+  - `ApprovalVotingMessage::CheckAndImportAssignment`
+  - `ApprovalVotingMessage::CheckAndImportApproval`
+  - `ApprovalVotingMessage::ApprovedAncestor`
+
+Output:
+  - `ApprovalNetworkingMessage::DistributeAssignment`
+  - `ApprovalNetworkingMessage::DistributeApproval`
+  - `RuntimeApiMessage::Request`
+  - `ChainApiMessage`
+  - `AvailabilityRecoveryMessage::Recover`
+  - `CandidateExecutionMessage::ValidateFromExhaustive`
+
+## Functionality
+
+The approval voting subsystem is responsible for casting votes and determining approval of candidates and as a result, blocks.
+
+This subsystem wraps a database which is used to store metadata about unfinalized blocks and the candidates within them. Candidates may appear in multiple blocks, and assignment criteria are chosen differently based on the hash of the block they appear in.
+
+## Database Schema
+
+The database schema is designed with the following goals in mind:
+  1. To provide an easy index from unfinalized blocks to candidates
+  1. To provide a lookup from candidate hash to approval status
+  1. To be easy to clear on start-up. What has happened while we were offline is unimportant.
+  1. To be fast to clear entries outdated by finality
+
+Structs:
+
+```rust
+struct TrancheEntry {
+    tranche: DelayTranche,
+    // assigned validators who have not yet approved, and the instant we received
+    // their assignment.
+    assignments: Vec<(ValidatorIndex, Tick)>,
+}
+
+struct OurAssignment {
+  cert: AssignmentCert,
+  tranche: DelayTranche,
+  validator_index: ValidatorIndex,
+  triggered: bool,
+}
+
+struct ApprovalEntry {
+    tranches: Vec<TrancheEntry>, // sorted ascending by tranche number.
+    backing_group: GroupIndex,
+    // When the next wakeup for this entry should occur. This is either to
+    // check a no-show or to check if we need to broadcast an assignment.
+    next_wakeup: Tick,
```

Comment:
Should we not rather make this a push driven system with timeouts? That usually reduces complexity in the implementation.

Reply:
I find it hard to justify introducing ticks (got to re-read it, I might have missed some details) - we are essentially implementing a coarse polling when futures and channels are available.

Reply:
How would you do it instead?

Reply:
The main thing I'm working around here is that we are dealing with these data:
  * Potentially many unfinalized blocks, forked with each other
  * Many candidates within those blocks, potentially with the same candidate appearing in multiple blocks
  * Assignments appearing for specific candidates in the context of particular blocks

For any particular `ApprovalEntry` we only want to poll the _first_ assignment that would time out as a no-show and not have a wakeup for each one.

Reply:
I suppose we could rewrite each `ApprovalEntry` to be its own `Future`, but the other thing I'm juggling is that we want this stuff to be persisted to disk as well, and we should prefer to write from a single task

Reply:
Writing from a single task is either using a [`FuturesUnordered`](https://docs.rs/futures/0.3.7/futures/stream/struct.FuturesUnordered.html) which then guarantees that we are on the same task assuming either chaining of futures or sync writes. We could also use a write queue if that is a concern.

Iiuc `ApprovalEntry` could impl `Future` iteself, which then lives in a `ApprovalEntrySet` which acts like a timerwheel and consume the `rx` end of `mpsc` for additional entries or implement `Sink` to consume additional `Entry` items.

Reply:
No matter what, we still need the `Tick` because that's our basic unit of time for this protocol. But I will explore different architectures. I still think that what's laid out in this doc sounds easiest to implement from an ownership perspective.